demos/common/mqtt_demo_helpers/mqtt_demo_helpers.c

```diff
@@ -407,6 +480,146 @@ static TransportSocketStatus_t prvConnectToServerWithBackoffRetries( NetworkCont
+         * will expect all the messages it sends to the broker to be sent back to it
+         * from the broker. This demo uses QOS0 in Subscribe, therefore, the Publish
+         * messages received from the broker will have QOS0. */
+        LogInfo( ( "Attempt to subscribe to the MQTT topic %s.", pcTopicFilter ) );
+        xResult = MQTT_Subscribe( pxMqttContext,
+                                  &xMQTTSubscription,
+                                  sizeof( xMQTTSubscription ) / sizeof( MQTTSubscribeInfo_t ),
+                                  globalSubscribePacketIdentifier );
+
+        if( xResult != MQTTSuccess )
+        {
+            LogError( ( "Failed to SUBSCRIBE to MQTT topic %s. Error=%s",
+                        pcTopicFilter, usTopicFilterLength ) );
+        }
+        else
+        {
+            xStatus = pdPASS;
+            LogInfo( ( "SUBSCRIBE sent for topic %s to broker.", pcTopicFilter ) );
+
+            /* Process incoming packet from the broker. After sending the subscribe, the
+             * client may receive a publish before it receives a subscribe ack. Therefore,
+             * call generic incoming packet processing function. Since this demo is
+             * subscribing to the topic to which no one is publishing, probability of
+             * receiving Publish message before subscribe ack is zero; but application
+             * must be ready to receive any packet.  This demo uses the generic packet
+             * processing function everywhere to highlight this fact. */
+            xResult = prvWaitForPacket( pxMqttContext, MQTT_PACKET_TYPE_SUBACK );
+
+            if( xResult != MQTTSuccess )
+            {
+                xStatus = pdFAIL;
+            }
+        }
+
+        if( xStatus == pdPASS )
+        {
+            /* Reset flag before checking suback responses. */
+            xFailedSubscribeToTopic = pdFALSE;
+
+            /* Check if recent subscription request has been rejected. #xTopicFilterContext is updated
+             * in the event callback to reflect the status of the SUBACK sent by the broker. It represents
+             * either the QoS level granted by the server upon subscription, or acknowledgement of
+             * server rejection of the subscription request. */
+            if( xTopicFilterContext.xSubAckStatus == MQTTSubAckFailure )
+            {
+                xFailedSubscribeToTopic = pdTRUE;
+
+                /* As the subscribe attempt failed, we will retry the connection after an
+                 * exponential backoff with jitter delay. */
+
+                /* Retry subscribe after exponential back-off. */
+                LogWarn( ( "Server rejected subscription request. Attempting to re-subscribe to topic %s.",
+                           xTopicFilterContext.pcTopicFilter ) );
+
+                xBackoffStatus = prvBackoffForRetry( &xRetryParams );
+                break;
```

Comment:
Is this break needed?

Reply:
Thanks for catching. This used to be inside a loop when there could have possibly been multiple topic filters.